### ISSUE TITLE
Absolute paths on Windows don't resolve after mapping

### ISSUE THREAD

**Issue by gadfly16** [LucaScheller/VFX-UsdAssetResolver#6]
Hi Luca, Thanks for the amazing work!

I'm working on the asset pipeline of our studio, that runs on Windows unfortunately. We would like to build our asset resolution logic around your `FileResolver`, but it seems to behave strangely in my (corner) case. If this behavior is by design, just let me know!

So my problem is that I would like to map absolute paths to absolute paths, something like this:

`X:/project/assets/Something.usd -> X:/project/assets/Something_v3.usd`

If I resolve the *target* directly it resolves correctly:

```
>>> stage.ResolveIdentifierToEditTarget("X:/project/assets/Something_v3.usd")
x:\project\assets\Something_v3.usd
``` 

But if this absolute path comes from a mapping transformation then it does not:

```
>>> fileResolver_context.GetMappingPairs()
{'X:/project/assets/Something.usd': 'X:/project/assets/Something_v3.usd'}
>>> stage.ResolveIdentifierToEditTarget("X:/project/assets/Something.usd")
   
```

The result is empty, as if the target wouldn't exist or something. Maybe I'm wrong but it seems to me that the resolver believes that the result of a mapping transformation **must** be relative, so it has to be looked up in `AR_SEARCH_PATHS`. But that's just my assumption.

Again, if it's by design just let me know!

Thanks,
Mate

**Comment by LucaScheller** [comment]
Hey Mate,
thanks for checking out the repo. 

Yes it is by design that only paths not starting with "./","../","/" are mapped via the mapping pairs.

Is there are reason why you'd want absolute paths? The benefit of only mapping paths that are on the search path is mainly speed, as the search path lookup can be skipped for absolute paths.
So while we could add the support, I'd prefer to keep it as is, as this is how the default resolver also does it.

Cheers,
Luca

**Comment by gadfly16** [comment]
Hey Luca,

I'm afraid my previous comment wasn't really on the spot. Let me try again.

Our shot building in Houdini is implemented in two custom nodes. One is called USD Writer, which is a wrapped up USD ROP node that detects if the output filename has a version number, and if so, it records the current version in a json file. Our other custom node is called Shot Def, which is a pimped up sublayer node, that reads (and writes) the shot's layers from a json file. 

My idea was that I extend the Shot Def node to let it build the mapping pairs from these json files, so any other downstream node referencing a layer could use the versionless names which would resolve to the current published version.

The problem is that if an artist uses an abolute path downstream (possibly because an environment variable resolves to an abolute path) the version resolution doesn't work. In our pipeline it's customary to load an asset like `$ASSETS/items/ball.usd` for example.

I hope this explanation makes more sense.

Thanks,
M

**Comment by gadfly16** [comment]
Hey Luca,

Finally we implemented our pipeline tools with the current design of your File Resolver. During implementation I developed a feeling that my need to resolve everything doesn't fit really well to the current design, so I wouldn't be surprised if you would have some hesitation wanting include this option. Good news is that we are good now, so there is no need to include it if you don't want to. Please free to close this issue.

One other thing, I would like to ask. In Solaris I buid my mapping pairs in a Python Script node and everythong works fine, but after rebuilding the mappings pairs I need to switch to another node and back to flush the USD cache. Simply recooking the downstream nodes do not reflect the changes in the mappings. Do you know a way how I can flush the cache programatically so all downstream nodes are forced to re-resolve and reload the layers?

Thanks for everything,
Mate

**Comment by LucaScheller** [comment]
Hey,
sorry for the slow response time, I'm currently switching jobs/relocating to another country and haven't had a lot of time for personal projects.

I still plan on adding the feature, some time in the next few months.

Are you refreshing the context as described here [Asset Resolver Docs](https://lucascheller.github.io/VFX-UsdAssetResolver/resolvers/FileResolver/PythonAPI.html#refreshing-the-resolver-context) ?

As far as I know,  there is no mechanism to force recook all affected nodes. You'll have to force cook them yourself (node.cook(force=True)).

Cheers,
Luca

**Comment by gadfly16** [comment]
No problem, and thank you for your answer!

Yes I'm refreshing the context. Force cooking and reloading the layers doesn't do the trick. The problem is placed much deeper, in the USD library's notorious cache subsystem. It seems that since it doesn't detect or doesn't care about the state change in the resolve context so it doesn't see the need to reload the layers. The ugly workaround is that you put the display flag on a node in a completely different graph (which means a different stage) and switch it back again. Then it re-resolves and reloads everything. (I suppose because of the stage change, but not sure) I thought there may be a way to tell the USD lib to release eveything from its cache, but I had no luck yet in the forums neither.

(By the way if you force recook any upstream node that will (force) recook all downstream nodes from there, since Houdini will mark every downstram node dirty in the DAG.)

Good luck with your new job!
Mate

**Comment by LucaScheller** [comment]
Hey, I added the absolute path exposing as an opt-in feature on the CachedResolver and FileResolver.

Release: https://github.com/LucaScheller/VFX-UsdAssetResolver/releases/tag/v0.6.6
Docs:
* FileResolver: https://lucascheller.github.io/VFX-UsdAssetResolver/resolvers/FileResolver/PythonAPI.html#resolver
* CachedResolver: https://lucascheller.github.io/VFX-UsdAssetResolver/resolvers/CachedResolver/overview.html#admonition-pro-tip-1

Let me know if it works!
Cheers,
Luca

**Comment by gadfly16** [comment]
Hey,
Thank you very much! Since then, we implemented our pipeline with the original design of the FileResolver. It is now in production, and everybody is pleased with the outcome. Because of this (and for some other reasons) in the near future I won't be able to test the new option unfortunately, but I hope in the next iteration we will be able to give it a go.
Thanks again, your work made this possible for us.
Best regards,
Mate

**Comment by LucaScheller** [comment]
No worries, I'll close the issue for now. Please re-open it once you have had time to test it and if there are any issues remaining.
